README.rst
================
datetime_periods
================

.. image:: https://travis-ci.org/mediapop/datetime_periods.png
           :target: https://travis-ci.org/mediapop/datetime_periods

This module aims to help you create time periods as easily as if by a
snap of your finger.

Pass in a `datetime.datetime()` object and a period name and it'll
return the beginning and end of that period.

Documentation available on `Read the Docs`_.

Installation:
-------------

You can install from pypi!

.. code-block::

    pip install datetime_periods


`period` usage
------

Pass in a `datetime.datetime()` object and a period name and it'll
return the beginning and end of that period.

.. code-block::

    >>> from datetime_periods import period
    >>> period(datetime(2012, 4, 2, second=12), 'minute')
    [datetime(2012, 4, 2), datetime(2012, 4, 2, 0, 0, 59)]
    >>> period(datetime(2012, 4, 2), 'hour')
    [datetime(2012, 4, 2, 0), datetime(2012, 4, 2, 0, 59, 59)]
    >>> period(datetime(2012, 4, 2), 'day')
    [datetime(2012, 4, 2), datetime(2012, 4, 2, 23, 59, 59)]
    >>> period(datetime(2012, 4, 2), 'week')
    [datetime(2012, 4, 2), datetime(2012, 4, 8, 23, 59, 59)]
    >>> period(datetime(2012, 4, 15), 'month')
    [datetime(2012, 4, 1), datetime(2012, 4, 30, 23, 59, 59)]
    >>> period(datetime(2012, 4, 2), 'quarter')
    [datetime(2012, 4, 1), datetime(2012, 6, 30, 23, 59, 59)]
    >>> period(datetime(2012, 9, 1), 'half_year')
    [datetime(2012, 7, 1), datetime(2012, 12, 31, 23, 59, 59)]
    >>> period(datetime(2012, 7, 1), 'year')
    [datetime(2012, 1, 1), datetime(2012, 12, 31, 23, 59, 59)]


`TimeRange` usage
------

The `TimeRange` class takes two times, `start` and `stop`, and creates
`datetime` objects from them that is smart about when a date should
roll over to the following day.

This class can also act like a 2 length list where index 0=start,
1=stop time. This to allow the class to be used for argument
expansion and as an iterator.

.. code-block::

    >>> from datetime_periods import TimeRange
    >>> tr = TimeRange('17:00', '23:00', '2013-12-25')
    >>> tr.start
    datetime(2013, 12, 25, 17)
    >>> tr.stop
    datetime(2013, 12, 25, 23)
    >>> tr = TimeRange('17:00', '04:00', '2013-12-25')
    >>> tr.start
    datetime(2013, 12, 25, 17)
    >>> tr.stop
    datetime(2013, 12, 26, 4)
    >>> tr[0] == tr.start
    True
    >>> tr[1] == tr.stop
    True

Sugar
-----

The `sugar` module has sugar functions for all variants available.

Sugar functions for entire period:

* `period_second`
* `period_minute`
* `period_hour`
* `period_day`
* `period_week`
* `period_month`
* `period_quarter`
* `period_half_year`
* `period_year`

Sugar functions for beginning of period:

* `period_beginning_second`
* `period_beginning_minute`
* `period_beginning_hour`
* `period_beginning_day`
* `period_beginning_week`
* `period_beginning_month`
* `period_beginning_quarter`
* `period_beginning_half_year`
* `period_beginning_year`

Sugar functions for end of period:

* `period_end_second`
* `period_end_minute`
* `period_end_hour`
* `period_end_day`
* `period_end_week`
* `period_end_month`
* `period_end_quarter`
* `period_end_half_year`
* `period_end_year`

.. _Read the Docs: http://datetime_periods.readthedocs.org/en/latest/


CHANGELOG
2013-12-28  Björn Andersson  <ba@sanitarium.se>

	* datetime_periods/time_range.py (TimeRange): Add TimeRange class
	to smartly handle ranges of time. Main example is opening hours a
	location opens at 17:00 and closes at 03:00, it closes at another
	date.


tests/test_sugar.py
from datetime import datetime
import unittest

from datetime_periods import sugar


class TestSugarFunctions(unittest.TestCase):
    def test_all_of_them(self):
        '''
        Since the sugar methods have all been tested on their own the logic
        of them should be sound. But to ensure that a typo made it so that
        they don't work at all let's just iterate over them and send in a
        datetime to see if no exceptions are raised.
        '''
        test_datetime = datetime(2012, 3, 1)
        for function in dir(sugar):
            if function.startswith('_'): continue

            getattr(sugar, function)(test_datetime)


tests/test_period.py
from datetime import datetime
import unittest

from datetime_periods import period


class TestPeriod(unittest.TestCase):
    def test_period_second(self):
        beginning = datetime(2012, 4, 1, microsecond=5000)
        self.assertEqual(period(beginning, 'second'),
                         [datetime(2012, 4, 1, microsecond=0),
                          datetime(2012, 4, 1, microsecond=999999)])

    def test_period_minute(self):
        beginning = datetime(2012, 4, 2, second=12)
        self.assertEqual(period(beginning, 'minute'),
                         [datetime(2012, 4, 2),
                          datetime(2012, 4, 2, 0, 0, 59)])

    def test_period_hour(self):
        beginning = datetime(2012, 4, 2, minute=20)
        self.assertEqual(period(beginning, 'hour'),
                         [datetime(2012, 4, 2),
                          datetime(2012, 4, 2, 0, 59, 59)])

    def test_period_day(self):
        beginning = datetime(2012, 4, 2, hour=12)
        self.assertEqual(period(beginning, 'day'),
                         [datetime(2012, 4, 2),
                          datetime(2012, 4, 2, 23, 59, 59)])

    def test_period_week(self):
        beginning = datetime(2012, 4, 3)
        self.assertEqual(period(beginning, 'week'),
                         [datetime(2012, 4, 2),
                          datetime(2012, 4, 8, 23, 59, 59)])

    def test_period_month(self):
        beginning = datetime(2012, 4, 15)
        self.assertEqual(period(beginning, 'month'),
                         [datetime(2012, 4, 1),
                          datetime(2012, 4, 30, 23, 59, 59)])

    def test_period_quarter(self):
        beginning = datetime(2012, 5, 15)
        self.assertEqual(period(beginning, 'quarter'),
                         [datetime(2012, 4, 1),
                          datetime(2012, 6, 30, 23, 59, 59)])

    def test_period_half_year(self):
        beginning = datetime(2012, 9, 1)
        self.assertEqual(period(beginning, 'half_year'),
                         [datetime(2012, 7, 1),
                          datetime(2012, 12, 31, 23, 59, 59)])

    def test_period_year(self):
        beginning = datetime(2012, 7, 1)
        self.assertEqual(period(beginning, 'year'),
                         [datetime(2012, 1, 1),
                          datetime(2012, 12, 31, 23, 59, 59)])


tests/test_period_end.py
from datetime import datetime
import unittest

from datetime_periods import period_end


class TestPeriodEnd(unittest.TestCase):
    def test_period_end_second(self):
        beginning = datetime(2012, 4, 2)
        self.assertEqual(period_end(beginning, 'second'),
                         datetime(2012, 4, 2, 0, 0, 0, 999999))

    def test_period_end_minute(self):
        beginning = datetime(2012, 4, 2)
        self.assertEqual(period_end(beginning, 'minute'),
                         datetime(2012, 4, 2, 0, 0, 59))

    def test_period_end_hour(self):
        beginning = datetime(2012, 4, 2)
        self.assertEqual(period_end(beginning, 'hour'),
                         datetime(2012, 4, 2, 0, 59, 59))

    def test_period_end_day(self):
        beginning = datetime(2012, 4, 2)
        self.assertEqual(period_end(beginning, 'day'),
                         datetime(2012, 4, 2, 23, 59, 59))

    def test_period_end_week(self):
        beginning = datetime(2012, 4, 2)
        self.assertEqual(period_end(beginning, 'week'),
                         datetime(2012, 4, 8, 23, 59, 59))

    def test_period_end_month(self):
        beginning = datetime(2012, 4, 1)
        self.assertEqual(period_end(beginning, 'month'),
                         datetime(2012, 4, 30, 23, 59, 59))

    def test_period_end_quarter(self):
        beginning = datetime(2012, 4, 1)
        self.assertEqual(period_end(beginning, 'quarter'),
                         datetime(2012, 6, 30, 23, 59, 59))

    def test_period_end_half_year(self):
        beginning = datetime(2012, 7, 1)
        self.assertEqual(period_end(beginning, 'half_year'),
                         datetime(2012, 12, 31, 23, 59, 59))

    def test_period_end_year(self):
        beginning = datetime(2012, 1, 1)
        self.assertEqual(period_end(beginning, 'year'),
                         datetime(2012, 12, 31, 23, 59, 59))


datetime_periods/sugar.py
from datetime_truncate import truncate

from .period import period
from .period_end import period_end


def period_second(datetime):
    ''' Sugar for :py:func:`period(datetime, 'second')` '''
    return period(datetime, 'second')


def period_minute(datetime):
    ''' Sugar for :py:func:`period(datetime, 'minute')` '''
    return period(datetime, 'minute')


def period_hour(datetime):
    ''' Sugar for :py:func:`period(datetime, 'hour')` '''
    return period(datetime, 'hour')


def period_day(datetime):
    ''' Sugar for :py:func:`period(datetime, 'day')` '''
    return period(datetime, 'day')


def period_week(datetime):
    ''' Sugar for :py:func:`period(datetime, 'week')` '''
    return period(datetime, 'week')


def period_month(datetime):
    ''' Sugar for :py:func:`period(datetime, 'month')` '''
    return period(datetime, 'month')


def period_quarter(datetime):
    ''' Sugar for :py:func:`period(datetime, 'quarter')` '''
    return period(datetime, 'quarter')


def period_half_year(datetime):
    ''' Sugar for :py:func:`period(datetime, 'half')` '''
    return period(datetime, 'half_year')


def period_year(datetime):
    ''' Sugar for :py:func:`period(datetime, 'year')` '''
    return period(datetime, 'year')


def period_end_second(datetime):
    ''' Sugar for :py:func:`period_end(datetime, 'second')` '''
    return period_end(datetime, 'second')


def period_end_minute(datetime):
    ''' Sugar for :py:func:`period_end(datetime, 'minute')` '''
    return period_end(datetime, 'minute')


def period_end_hour(datetime):
    ''' Sugar for :py:func:`period_end(datetime, 'hour')` '''
    return period_end(datetime, 'hour')


def period_end_day(datetime):
    ''' Sugar for :py:func:`period_end(datetime, 'day')` '''
    return period_end(datetime, 'day')


def period_end_week(datetime):
    ''' Sugar for :py:func:`period_end(datetime, 'week')` '''
    return period_end(datetime, 'week')


def period_end_month(datetime):
    ''' Sugar for :py:func:`period_end(datetime, 'month')` '''
    return period_end(datetime, 'month')


def period_end_quarter(datetime):
    ''' Sugar for :py:func:`period_end(datetime, 'quarter')` '''
    return period_end(datetime, 'quarter')


def period_end_half_year(datetime):
    ''' Sugar for :py:func:`period_end(datetime, 'half')` '''
    return period_end(datetime, 'half_year')


def period_end_year(datetime):
    ''' Sugar for :py:func:`period_end(datetime, 'year')` '''
    return period_end(datetime, 'year')


def period_beginning_second(datetime):
    ''' Sugar for :py:func:`datetime_truncate.truncate(datetime, 'second')` '''
    return truncate(datetime, 'second')


def period_beginning_minute(datetime):
    ''' Sugar for :py:func:`datetime_truncate.truncate(datetime, 'minute')` '''
    return truncate(datetime, 'minute')


def period_beginning_hour(datetime):
    ''' Sugar for :py:func:`datetime_truncate.truncate(datetime, 'hour')` '''
    return truncate(datetime, 'hour')


def period_beginning_day(datetime):
    ''' Sugar for :py:func:`datetime_truncate.truncate(datetime, 'day')` '''
    return truncate(datetime, 'day')


def period_beginning_week(datetime):
    ''' Sugar for :py:func:`datetime_truncate.truncate(datetime, 'week')` '''
    return truncate(datetime, 'week')


def period_beginning_month(datetime):
    ''' Sugar for :py:func:`datetime_truncate.truncate(datetime, 'month')` '''
    return truncate(datetime, 'month')


def period_beginning_quarter(datetime):
    '''
    Sugar for :py:func:`datetime_truncate.truncate(datetime, 'quarter')`
    '''
    return truncate(datetime, 'quarter')


def period_beginning_half_year(datetime):
    ''' Sugar for :py:func:`datetime_truncate.truncate(datetime, 'half')` '''
    return truncate(datetime, 'half_year')


def period_beginning_year(datetime):
    ''' Sugar for :py:func:`datetime_truncate.truncate(datetime, 'year')` '''
    return truncate(datetime, 'year')


datetime_periods/period.py
from datetime_truncate import truncate as period_beginning
from .period_end import period_end


def period(datetime, period_name='day'):
    '''
    Takes the given `datetime` and then creates the `period_name` that
    `datetime` belongs to. If given one in the middle of the day and
    `period_name` 'day' then it'll be from 00:00:00 till 23:59:59.

    Possible values for `period_name`:

    * second
    * minute
    * hour
    * day
    * week (iso week i.e. monday to sunday)
    * month
    * quarter
    * half_year
    * year

    Examples::

        >>> period(datetime(2012, 4, 2), 'hour')
        [datetime(2012, 4, 2, 0), datetime(2012, 4, 2, 0, 59, 59)]
        >>> period(datetime(2012, 4, 2), 'day')
        [datetime(2012, 4, 2), datetime(2012, 4, 2, 23, 59, 59)]
        >>> period(datetime(2012, 4, 2), 'week')
        [datetime(2012, 4, 2), datetime(2012, 4, 8, 23, 59, 59)]
        >>> period(datetime(2012, 4, 2), 'quarter')
        [datetime(2012, 4, 1), datetime(2012, 6, 30, 23, 59, 59)]

    :params datetime: A truncated datetime object
    :params period_name: The period for which to calculate the end for
                         `datetime`
    :return: datetime with all fields to second set to the very last before
             before the next period
    :rtype: :py:mod:`datetime` datetime object
    '''
    start = period_beginning(datetime, period_name)
    return [start,
            period_end(start, period_name)]


datetime_periods/time_range.py
from datetime import datetime, timedelta

from dateutil.parser import parse
import six


class TimeRange(object):
    """Takes two times, ``start`` and ``stop``, and tries to be
    smart about putting a date to those times.

    Stop is always assumed to follow start in chronological order, so
    if stop is numerically less than start then it must be tomorrow.

    Strings are parsed with ``dateutil.parser`` from the
    `python-dateutil` package.

    This class can also act like a 2 length list where index 0=start,
    1=stop time. This to allow the class to be used for argument
    expansion and as an iterator.

    Examples::

        >>> tr = TimeRange('17:00', '23:00', '2013-12-25')
        >>> tr.start
        datetime(2013, 12, 25, 17)
        >>> tr.stop
        datetime(2013, 12, 25, 23)
        >>> tr = TimeRange('17:00', '04:00', '2013-12-25')
        >>> tr.start
        datetime(2013, 12, 25, 17)
        >>> tr.stop
        datetime(2013, 12, 26, 4)
        >>> tr[0] == tr.start
        True
        >>> tr[1] == tr.stop
        True

    :param start: a ``time`` object or a time string
    :param stop: a ``time`` object or a time string
    :param date: a ``datetime``, ``date``, or date string
    :param tzinfo: None or a tzinfo that will replace the current one
                   in the ``timestamp`` attributes ``start`` and ``stop``

    """
    def __init__(self, start, stop, date=None, tzinfo=None):
        self.date = self._ensure_date(date)
        self.tzinfo = tzinfo

        self.time = {
            'start': self._ensure_time(start),
            'stop': self._ensure_time(stop),
        }

        self._timestamps()

    def _timestamps(self):
        self.timestamp = {}
        start, stop = self.time['start'], self.time['stop']

        self.timestamp['start'] = datetime.combine(self.date, start)
        self.timestamp['stop'] = datetime.combine(self.date, stop)

        if stop <= start:
            self.timestamp['stop'] = self.timestamp['stop'] + timedelta(days=1)

        if self.tzinfo:
            for k, v in self.timestamp.items():
                self.timestamp[k] = v.replace(tzinfo=self.tzinfo)

    def _ensure_date(self, date):
        if not date:
            date = datetime.now()
        elif isinstance(date, six.string_types):
            date = parse(date, yearfirst=True)  # ISO8601 dates ftw

        try:
            return date.date()
        except AttributeError:
            return date

    def _ensure_time(self, timestamp):
        if isinstance(timestamp, six.string_types):
            timestamp = parse(timestamp)

        try:
            return timestamp.time()
        except AttributeError:
            return timestamp

    def __getitem__(self, key):
        if key == 0:
            return self.start
        elif key == 1:
            return self.stop
        else:
            raise KeyError

    def __iter__(self):
        return iter((self.start, self.stop))

    def __repr__(self):
        return('{cls}(start={start}, stop={stop}, '
               'date={date}, tzinfo={tzinfo})').format(
            cls=self.__class__.__name__,
            start=self.time['start'].__repr__(),
            stop=self.time['stop'].__repr__(),
            date=self.date.__repr__(),
            tzinfo=self.tzinfo.__repr__(),
        )

    @property
    def start(self):
        """The inserted ``start`` as a timestamp, if ``tzinfo`` was passed in
        it'll be set
        """
        return self.timestamp['start']

    @property
    def stop(self):
        """The inserted ``stop`` as a timestamp, if ``tzinfo`` was passed in
        it'll be set
        """
        return self.timestamp['stop']


datetime_periods/period_end.py
from dateutil.relativedelta import relativedelta

PERIODS = {
    'second': dict(seconds=1, microseconds=-1),
    'minute': dict(minutes=1, seconds=-1),
    'hour': dict(hours=1, seconds=-1),
    'day': dict(days=1, seconds=-1),
    'week': dict(weeks=1, seconds=-1),
    'month': dict(months=1, seconds=-1),
    'quarter': dict(months=3, seconds=-1),
    'half_year': dict(months=6, seconds=-1),
    'year': dict(years=1, seconds=-1),
}


def period_end(datetime, period_name='day'):
    '''
    Returns a datetime where it is the end of `period_name`.
    Notice that :py:func:`period_end()` assumes that the `datetime` has
    been run by :py:func:`truncate()` before being passed in. If that
    is not the case the results might not be what is expected.

    Possible values for `period_name`:

    * second
    * minute
    * hour
    * day
    * week (iso week i.e. monday to sunday)
    * month
    * quarter
    * half_year
    * year

    Examples::

        >>> period_end(datetime(2012, 4, 2), 'hour')
        datetime(2012, 4, 2, 0, 59, 59)
        >>> period_end(datetime(2012, 4, 2), 'day')
        datetime(2012, 4, 2, 23, 59, 59)
        >>> period_end(datetime(2012, 4, 2), 'week')
        datetime(2012, 4, 8, 23, 59, 59)
        >>> period_end(datetime(2012, 4, 1), 'quarter')
        datetime(2012, 6, 30, 23, 59, 59)

    :params datetime: A truncated datetime object
    :params period_name: The period for which to calculate the end for
                         `datetime`
    :return: datetime with all fields to second set to the very last before
             before the next period
    :rtype: :py:mod:`datetime` datetime object
    '''
    if period_name in PERIODS:
        return datetime + relativedelta(**PERIODS[period_name])
    else:
        raise ValueError('{} is not a valid period_name. Valid: {}'.format(
            period_name,
            ', '.join(PERIODS.keys())
        ))
